File: framework/scala/src/main/scala/soda/scala/web/resp/JsonResponse.scala
package soda.scala.web.resp

import play.api.libs.json._

import java.nio.charset.StandardCharsets

class JsonResponse(httpCode: Int, var body: JsValue)
  extends BaseResponse(httpCode, "application/json; charset=utf-8") {

  override def getBody: Array[Byte] = {
    Json.stringify(body).getBytes(StandardCharsets.UTF_8)
  }

}

object JsonResponse {
  def create[T](httpCode: Int, body: T)(implicit fmt: Writes[T]): JsonResponse = {
    new JsonResponse(httpCode, Json.toJson(body))
  }
}

File: framework/kotlin/src/main/kotlin/soda/kotlin/web/resp/JsonResponse.kt
package soda.kotlin.web.resp

import kotlinx.serialization.KSerializer
import kotlinx.serialization.serializer
import kotlinx.serialization.json.Json
import kotlin.reflect.typeOf
import java.nio.charset.StandardCharsets

class JsonResponse(httpCode: Int, private val body: String):
    BaseResponse(httpCode, "application/json; charset=utf-8") {

    override fun getBody(): ByteArray {
        return body.toByteArray(StandardCharsets.UTF_8)
    }

    companion object {

        inline fun <reified T> create(httpCode: Int, body: T): JsonResponse {
            @Suppress("UNCHECKED_CAST")
            val ks = serializer(typeOf<T>()) as KSerializer<T>
            return JsonResponse(httpCode, Json.encodeToString(ks, body))
        }

    }

}